### PR TITLE
fix: commit watch txm and reply empty array on empty EXEC

### LINKED ISSUE #506: WATCH then empty EXEC leaks the watch transaction (and returns nil instead of empty array)
`WATCH key; MULTI; EXEC` with no queued commands leaks the RepeatableRead watch transaction: `MultiExec` early-returns on an empty command list without committing or aborting the txm, and the handler then nulls `txm_` so its destructor's abort is skipped. The txm and its read locks leak.

## Evidence

`WatchKeys` assigns the watch txm (`src/redis_handler.cpp:1877-1884`):
```cpp
if (txm_ == nullptr)
    txm_ = redis_impl_->NewTxm(txn_isolation_level_, txn_protocol_);  // RepeatableRead + OCC
```
`MultiExec` empty-list early return (`src/redis_service.cpp:2066-2069`):
```cpp
if (cmd_reqs.empty()) { redis_reply.OnNil(); return NO_ERROR; }   // txm untouched
```
`MultiTransactionHandler::Run` nulls `txm_` after the call (`redis_handler.cpp:1829`), so the destructor's `if (txm_) AbortTx(txm_)` (`:1680`) is a no-op → the watch txm is never aborted, leaking it (and any read locks/intents it holds) from the pool.

## Repro
`WATCH k` then `MULTI` then `EXEC` (empty queue) → leaks one txm; also replies `nil` whereas Redis replies an empty array `(empty array)`.

**Fix:** in the empty-EXEC path, abort (or commit) the passed-in txm before returning; return an empty array to match Redis.

---
Found during a code audit (docs PR #492). Verified against source at the cited lines.

🤖 Found with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by liunyl**
## Summary

`WATCH k; MULTI; EXEC` with no queued commands leaked the RepeatableRead watch transaction and replied a RESP null instead of an empty array. This PR makes the empty-EXEC path consume the watch txm and aligns every aborted-EXEC reply with real Redis.

## Root cause

- `MultiExec` early-returned on an empty command list without committing or aborting the passed-in txm (`src/redis_service.cpp`).
- `MultiTransactionHandler::Run` then nulls `txm_` unconditionally, so the destructor's abort backstop never fires — the `TransactionExecution` is never returned to the pool.
- Additionally, every aborted-EXEC path replied a null **bulk string** (`$-1`) where Redis replies a null **array** (`*-1`), and an empty EXEC replied null instead of `*0`.

## Fix

- The empty-EXEC branch now **commits** the watch txm when one exists, so watched-key validation still runs: a broken watch fails the commit and EXEC replies `*-1` (Redis dirty-CAS semantics apply even to an empty queue); otherwise EXEC replies `*0`. `MultiExec` now consumes the txm on every path, restoring the invariant that every `NewTxm` pairs with `CommitTx`/`AbortTx`.
- All aborted-EXEC replies now emit a RESP null array via a new `RedisReplier::OnNullArray()` (exec-error abort, failed commit, poisoned pre-EXEC watch failure).
- The `WatchKeys with ISO=ReadCommitted` test enabled a fault injector that does not exist (`watch_keys_txm_iso_level`) and has been running on the default isolation level; it now uses the real injector (`txm_iso_level_read_committed`, guarded by `try/finally`) and also covers the empty-EXEC conflict, which works under ReadCommitted because watch-keys reads are promoted to RepeatableRead (`data_substrate/tx_service/src/tx_execution.cpp:7139-7143`).

## Wire behavior (verified with raw RESP probes)

| Scenario | Before | After | Redis |
|---|---|---|---|
| `MULTI; EXEC` (empty) | `$-1` | `*0` | `*0` |
| `WATCH k; MULTI; EXEC` (empty, unmodified) | `$-1` + **txm leak** | `*0` | `*0` |
| `WATCH k; k modified; MULTI; EXEC` (empty) | `$-1` + **txm leak** | `*-1` | `*-1` |
| `WATCH k; k modified; MULTI; PING; EXEC` | `$-1` | `*-1` | `*-1` |
| `MULTI; UNWATCH; EXEC` | `$-1` | `*0` | `*1` (`OK`) — see note |

Note: UNWATCH inside MULTI replies OK immediately in EloqKV instead of being queued (pre-existing divergence; Redis queues it and EXEC replies a one-element array). This change only moves that path's reply from `$-1` to `*0` — strictly closer to Redis — and a test pins the behavior so any future UNWATCH-queueing work must touch it consciously.

## Leak evidence (RSS of the server process, 50k-cycle pipelined batches)

| Workload | Before fix | After fix |
|---|---|---|
| 50k × `WATCH k; MULTI; EXEC` (batch 1) | **+677 MB** | +0.4 MB |
| 50k × `WATCH k; MULTI; EXEC` (batch 2) | **+690 MB** | +0.0 MB |
| 50k × `MULTI; PING; EXEC` (control) | +13 MB | (comparable) |

≈14 KB leaked per cycle before the fix, monotonic and never reclaimed. There is no exposed counter for live external txms (`ext_active_tx_cnt_` is private, no INFO/DEBUG/metrics surface), so the automated regression tests pin the reply semantics and repeatedly exercise the previously-leaking path; a TCL reclamation assertion needs new observability surface — tracked in #528.

## Tests

7 new raw-RESP regression tests in `tests/unit/eloq/multi.tcl` (the Tcl client decodes `*0`, `*-1`, `$-1` all as `{}`, so assertions use `readraw` via a small `exec_raw` helper), plus the repaired fault-injection test (Debug builds). Full `multi`/`multi2` suites pass on Release; `multi` incl. fault injection passes on Debug.

## Docs

`docs/02-command-processing.md` MULTI/EXEC and WATCH sections updated (empty-EXEC semantics, null-array replies, `EXECABORT` for queue-time errors, watch-read isolation promotion), per the repo doc-maintenance rule.

## Residual notes (reviewed, intentionally not changed here)

- `RedisReplier::OnNullArray()` is deliberately not part of the `OutputHandler` interface — only the RESP EXEC path emits it; Lua never does.
- The OCC retry loop's safety for watch txms rests on `WatchKeys` setting `tx_retrieable_ = false`; verified to hold on all current paths. A defensive assert could harden this for future edits.
- Engine-level exec-error aborts (`redis_service.cpp` exec-abort site) have no deterministic client-visible trigger without fault injection; the commit-failure site of the same shape is covered by tests.

Fixes #506

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Updated Redis RESP replies for WATCH/MULTI/EXEC to distinguish null array (`*-1`) from null bulk string.
  * Empty `EXEC` now returns `*0` (and commits an existing watch when applicable) instead of `nil`.
  * Aborted or failed `EXEC`/commit paths now return `*-1`.
* **Documentation**
  * Clarified MULTI/EXEC, WATCH, and OCC-retry semantics, including empty-queue EXEC and error reply behavior.
* **Tests**
  * Added regression tests for raw-wire `EXEC` outcomes and transactional watch/OCC edge cases.

### CODE REVIEW

File: tests/unit/eloq/multi.tcl [hunk 1]
```diff
@@ -230,6 +230,82 @@ start_server {tags {"multi"}} {
...
     } {OK}
 
+    # Run EXEC in readraw mode and return the raw first RESP line. The Tcl
+    # client decodes *0, *-1 and $-1 all as {}, so tests that pin the wire
+    # format of an EXEC reply must read it raw.
+    proc exec_raw {} {
+        r readraw 1
+        set res [r exec]
+        r readraw 0
+        return $res
```
[the remark this answers is not in the document]
Comment (liunyl) [reply tests/unit/eloq/multi.tcl:243]: Adopted in 1e21635 — `exec_raw` now restores raw mode in a `try/finally`. (For the record: in readraw mode the Tcl client returns error replies as raw lines without raising, so the only throwing path here is a socket I/O failure — but the guard is free and makes the helper unconditionally safe.)

**Review by thweetkomputer — APPROVED** [review]
